include/p4mlir/Dialect/P4HIR/P4HIR_Types.td

```diff
@@ -401,6 +401,27 @@ def HeaderType : StructLikeType<"Header", "header"> {
   }];
 }
 
+//===----------------------------------------------------------------------===//
+// HeaderUnionType
+//===----------------------------------------------------------------------===//
+
+// A header union. Declares the P4HIR::HeaderUnionType in C++.
+def HeaderUnionType : StructLikeType<"HeaderUnion", "header_union"> {
+  let summary = "header union type";
+  let description = [{
+    A header union represents an alternative containing at most one of several different headers.
+    !p4hir.header_union<"name", field1: HeaderType1, field2: HeaderType2>
+  }];
+
+  // We skip default builders entirely to consistently add validity bit field on fly
```

Comment:
Do we really need this? One can construct union from headers only that already contain validity bits.